File: python/104_maximum_depth_of_binary_tree.py
'''
104. Maximum Depth of Binary Tree
'''

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def maxDepthHelper(root, depth):
            if not root:
                return depth
            return max(maxDepthHelper(root.left, depth+1),
                       maxDepthHelper(root.right, depth+1))
        return maxDepthHelper(root, 0)

File: golang/104_maximum_depth_of_binary_tree.go
/*
104. Maximum Depth of Binary Tree
*/

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	maxLeft := maxDepth(root.Left)
	maxRight := maxDepth(root.Right)
	if maxLeft > maxRight {
		return maxLeft + 1
	}
	return maxRight + 1
}

